com.spotify/index.js

```diff
@@ -0,0 +1,662 @@
+const PLAY_URL = 'https://api.spotify.com/v1/me/player/play';
+
+// Stateless POST commands
+const NEXT_URL = 'https://api.spotify.com/v1/me/player/next';
+const PREVIOUS_URL = 'https://api.spotify.com/v1/me/player/previous';
+
+// Stateful PUT commands
+const SHUFFLE_URL = 'https://api.spotify.com/v1/me/player/shuffle';
+const REPEAT_URL = 'https://api.spotify.com/v1/me/player/repeat';
+const SEEK_URL = 'https://api.spotify.com/v1/me/player/seek';
+const TRACKS_URL = 'https://api.spotify.com/v1/me/tracks';
+
+// Search URL
+const SEARCH_URL = 'https://api.spotify.com/v1/search';
+
+//User Library URLs
+// need to replace get userId
+const USER_PLAYLISTS = 'https://api.spotify.com/v1/users/{username}/playlists';
+const PER_SET = 50;
+
+//Albums
+const ALBUM_URL = 'https://api.spotify.com/v1/albums/{id}/tracks';
+
+
+module.exports = class SpotifyDevice extends Tp.BaseDevice {
+  static get runOAuth2() {
+    return Tp.Helpers.OAuth2({
+      kind: 'com.spotify',
+      scope: [
+        'playlist-read-collaborative',
+        'playlist-modify-private',
+        'playlist-read-private',
+        'playlist-modify-public',
+        'user-read-email',
+        'user-read-private',
+        'user-read-birthdate',
+        'user-read-playback-state',
+        'user-read-currently-playing',
+        'user-modify-playback-state',
+        'user-read-recently-played',
+        'user-top-read',
+        'user-follow-read',
+        'user-follow-modify',
+        'user-library-read',
+        'user-library-modify'],
+      set_state: true,
+      authorize: 'https://accounts.spotify.com/authorize',
+      get_access_token: 'https://accounts.spotify.com/api/token',
+      redirect_uri: platform.getOrigin() + '/devices/oauth2/callback/com.spotify',
+
+      callback(engine, accessToken, refreshToken) {
+        return Tp.Helpers.Http.get(USER_URL, {
+          accept: 'application/json',
+          extraHeaders: {
+            'Authorization': 'Bearer ' + accessToken
+
+          }
+        }).then(response => {
+          const parsed = JSON.parse(response);
+          console.log(parsed);
+          return engine.devices.loadOneDevice(
+            {
+              kind: 'com.spotify',
+              accessToken: accessToken,
+              refreshToken: refreshToken,
+              userId: parsed.id,
+              userName: parsed.display_name
+            },
+            true
+          );
+        });
+      }
+    });
+  }
+
+  constructor(engine, state) {
+    super(engine, state);
+
+    this.uniqueId = 'com.spotify-' + this.userId;
+    this.name = 'Spotify Account of %s'.format(this.userName);
+    this.description = 'This is your Spotify Account';
+  }
+
+  get userId() {
+    return this.state.userId;
+  }
+
+  get userName() {
+    return this.state.userName;
+  }
+
+  get_get_user_name() {
+    return [{username: this.state.userName}];
+  }
+
+  get_get_user_id() {
+    return [{id: this.state.userId}];
+  }
```

Comment:
These two functions are not terribly useful, as the user is likely to know their name and it's displayed in My Almond anyway.
Perhaps they can be removed?

(having fewer functions improves NL quality)